\documentclass{beamer}

\usepackage{amsmath,amsthm,amssymb,enumerate}

\begin{document}

% what is dna
\begin{frame}
\frametitle{DNA Overview}
Deoxyribonucleic Acid (DNA) is
\begin{itemize}
\item Ubiquitous. Present in all known organisms.
\item Well suited to encoding strings over a finite alphabet.
\item Relatively stable and resilient.
\item Extremely small. Think nanometres. 
\item Modifiable. Interacts in interesting ways with specific proteins.
\end{itemize}
\ldots among many other things. This makes DNA interesting from a computational perspective. We like strings, and if DNA allows easy storage and manipulation of strings at nanoscale, we want to know about that.
\end{frame}

\begin{frame}
\frametitle{DNA Overview}
single strands and nucleotide sequences, complementary nucleotides, complementary sequences, 3' and 5' ends, and double strands
\end{frame}

% do we really need to go over why DNA encodes binary information?
%\begin{frame}
%\frametitle{Details: Base Pairs and Bits}
%\end{frame}

\begin{frame}
\frametitle{Things to do With DNA}
read it (sequencing), write/create it (how?), and PCR. (for example...)
\end{frame}

\begin{frame}
\frametitle{The Polymerase Chain Reaction Technique}
A \emph{DNA Polymerase} in an enzyme that, given a double strand of DNA in which one half is shorter than the other half, will use available nucleotides to extend a single strand as far as it can in the 5'-3' direction. 
\begin{center}
\includegraphics{dna_polymerase.png}
\end{center}
In the above image, the existing, shorter strand is coloured red.
\end{frame}

\begin{frame}
\frametitle{The Polymerase Chain Reaction Technique}
DNA polymerase will extend existing strands, but cannot create new strands on its own. A \emph{Primer} is a short, single strand of DNA that gives the polymerase something to extend. 
\begin{center}
\includegraphics{pcr_primer.png}
\end{center}
Note that since a nucleotide sequence will bind only to its complementary sequence, we can use DNA polymerase in combination with primers to complete only the single strands that contain the complementary sequence of some primer \emph{as a subsequence} to double strands. This is very powerful.
\end{frame}

\begin{frame}
\frametitle{The Polymerase Chain Reaction Technique}
Next, we note that the following happen in mutually exclusive temperature ranges:
\begin{itemize}
\item DNA Melting, in which a double strand separates into its component single strands ($\sim 95^\circ$ C).
\item DNA Annealing, in which nearby complementary sequences bind together, forming a double strand ($\sim 60^\circ$ C).
\item DNA Extension, in which DNA polymerase extends incomplete double strands ($\sim 75^\circ$ C).
\end{itemize}
By repeatedly changing the temperature of a solution comprised of DNA, DNA polymerase, and a variety of other things (what the polymerase needs to build the DNA, and so on), and adding the appropriate primers after each temperature change cycle, we can \emph{selectively duplicate} DNA strands at an exponential rate. This is referred to as a \emph{polymerase chain reaction} (PCR).
\end{frame}

\begin{frame}
\frametitle{The Polymerase Chain Reaction Technique}
Visually:
\includegraphics[width = \paperwidth]{polymerase_chain_reaction.jpg}
\end{frame}

% mention that PCR machines exist, show a picture of a modern one.

% logical operations on strings
\begin{frame}
Indeed, we can fairly reliably do the following to DNA (among other things):
amplify, separate, \emph{merge}, detect.
\end{frame}

% we encode characters in our finite alphabet as, say, a 30-mer, meaning 30 base pairs,
% to increase the effectiveness of the PCR-selection procedure. Mention this.

% assume error-free, then we have a model of computation! Introduce notation.

% the dna computing model (restricted? unrestricted?)
\begin{frame}
\frametitle{The Restricted Model}
A \emph{tube} is a multiset of strings over some alphabet $\Sigma$. We have the following operations on tubes:
\begin{itemize}
\item \emph{Separate}. Given a tube $T$ and a symbol $s \in \Sigma$, this produces two tubes $+(T,s)$ and $-(T,s)$ where $+(T,s)$ is the set of strings of $T$ which contain the symbol $s$, and $-(T,s)$ is the set of strings of $T$ which do not contain $s$. The original tube $T$ is destroyed.
\item \emph{Merge}. Given tubes $T_1,T_2$, this produces $T_1 \cup T_2$. $T_1, T_2$ are destroyed.
\item \emph{Detect}. Given a tube $T$, produces ``yes'' if $T$ is nonempty, and ``no'' otherwise.
\end{itemize}
\end{frame}

\begin{frame}

introduce program notation
\end{frame}

\begin{frame}
\frametitle{The Unrestricted Model}

The operations on the previous slide give what is called the \emph{restricted model} of DNA computing. The \emph{unrestricted model} is given by including the operation

\emph{Amplify}. Which, given a tube $T$, produces two new tubes $T_1, T_2$ such that $T_1 = T_2 = T$, and destroys $T$.
\end{frame}

% example: deciding whether or not a graph is 3-colourable (NP-complete problem)
\begin{frame}
\frametitle{Example: A DNA program for the 3-colourability problem}

Given an $n$ vertex graph $G$ with edges $e_1,\ldots,e_n$, let $\Sigma = \{r_1,b_1,g_1,r_2,b_2,g_2,\ldots,r_n,b_n,g_n\}$, and consider the following restricted program on input $T = \{\alpha | \alpha \subseteq \Sigma \wedge \alpha = c_1c_2,\ldots,c_n \wedge c_i \in \{r_i,b_i,c_i\}, i = 1,2,\ldots,n \}$

\begin{enumerate}[(1)]
\item Input$(T)$.
\item for each edge $e_k \in \{e_1,\ldots,e_z\}$, let $e_k = \langle i,j \rangle$ in
  \begin{enumerate}[(i)]
  \item $T_{red} \leftarrow +(T,r_i)$ and $T_{blue \vee green} \leftarrow -(T,r_i)$
  \item $T_{blue} \leftarrow +(T_{blue \vee green},b_i)$ and $T_{green} \leftarrow -(T_{blue \vee green},b_i)$
  \item $T_{red}^{good} \leftarrow -(T_{red},r_j)$
  \item $T_{blue}^{good} \leftarrow -(T_{blue},b_j)$
  \item $T_{green}^{good} \leftarrow -(T_{green},g_j)$
  \item $T' = T_{red}^{good} \cup T_{blue}^{good}$
  \item $T = T_{gree}^{good} \cup T'$
  \end{enumerate}
\item Output(Detect$(T)$).
\end{enumerate}
\end{frame}

% measures of computational complexity for dna computing
\begin{frame}
\frametitle{Measures of Complexity for DNA Programs}

The number of biological steps performed gives a straightforward measure of time complexity, and the highest number of strands of DNA present at any step during a DNA computaiton can be thought of as it's space complexity. Therefore, we consider the number of \emph{bio steps} a DNA program takes, and the number of \emph{strands} it requires, as a function of the input.

Note that in the restricted model, the number of strands required is number of strands in the input test tube, as no operation creates more strands.
\end{frame}

\begin{frame}
\frametitle{Example: Complexity of the program for 3-colourability}

For a graph with $n$ vertices and $k$ edges \ldots

The number of strands required is $\lvert T \rvert = 3^n$, one strand for each possible assignment of colours to vertices.

The number of bio steps required is $7k + 2 \in \mathcal{O}(k)$, as we perform steps $(i - vii)$ once for each edge, and a constant number of other steps. (Alternatively, $\mathcal{O}(n^2)$ as $k \leq n^2$).
\end{frame}

% table of results / complexity from Boneh "on the computational power of DNA"?

\begin{frame}

we seem to have solved an NP-complete problem (3-colourability) in time polynomial in the size of the graph!

(problems: the amount of DNA still grows exponentially, simulation of DNA computation on a TM does not preserve this due to the high degree of parallelism present in the separate operation)

\end{frame}

% the model is turing complete (winfree) 

\begin{frame}
It is easy to see that we can simulate both the restricted and unrestricted model with a Turing machine. (how to simulate other direction?) (winfree shows that only annealing and ligation are needed, must explain this earlier)

This is done by proving that we can simulate an arbitrary \emph{blocked cellular automaton} (BCA) with a DNA program. Since BCAs are capable of universal computation, so then are DNA programs.

\end{frame}

% do we really want to go in to this?

% complexity / BP / NBP (really?)

% error rates and the amplify operation

We have been assuming that our biological operations are error free. In reality, they are not. Further, DNA strands are prone to dissolving, and errors in the annealing process sometimes occur.

In the universal model, we can address this by using the amplify operation every so many steps to double the amount of DNA, and by double-encoding our strings in the starting test tube. (that is, instead of $s_1,\ldots,s_n$, encode $s_1,\ldots,s_n,s_1,\ldots,s_n$.

% real-world dna computing (adleman, the DES team, and impracticality)

% other biocomputing models 

Deoxyribozyme-based. (tictactoe)

Programmable automata that can interact with the surrounding microbiological environment.

\end{document}
